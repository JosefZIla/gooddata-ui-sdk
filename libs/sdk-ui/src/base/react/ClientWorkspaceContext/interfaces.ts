// (C) 2019-2022 GoodData Corporation

/**
 * Resolved LCM identifiers of the workspace.
 *
 * @alpha
 */
export interface IClientWorkspaceIdentifiers {
    /**
     * Data product identifier.
     */
    dataProduct?: string;

    /**
     * Client identifier.
     */
    client?: string;

    /**
     * Segment identifier.
     */
    segment?: string;

    /**
     * Workspace identifier.
     */
    workspace?: string;
}

/**
 * @internal
 */
export interface IClientWorkspaceStatus {
    isInitialized: boolean;
}
