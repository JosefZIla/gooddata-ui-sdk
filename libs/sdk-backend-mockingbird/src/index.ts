// (C) 2019-2021 GoodData Corporation
/**
 * @packageDocumentation
 */
import { dummyBackend, dummyBackendEmptyData, dummyDataView } from "@gooddata/sdk-backend-base";
export { dummyBackend, dummyBackendEmptyData, dummyDataView };

export { recordedBackend } from "./recordedBackend";

export {
    RecordingIndex,
    ExecutionRecording,
    InsightRecording,
    DisplayFormRecording,
    ScenarioRecording,
    RecordedBackendConfig,
    CatalogRecording,
    VisClassesRecording,
    DashboardRecording,
    RecordedRefType,
    SecuritySettingsUrlValidator,
    SecuritySettingsOrganizationScope,
} from "./recordedBackend/types";

export {
    recordedDataView,
    recordedDataViews,
    recordedInsight,
    DataViewAll,
    dataViewWindow,
    DataViewFirstPage,
    NamedDataView,
} from "./recordedBackend/execution";

export { CompositeBackendPart, compositeBackend } from "./compositeBackend/index";

export {
    LegacyExecutionRecording,
    legacyRecordedBackend,
    legacyRecordedDataView,
    LegacyRecordingIndex,
    LegacyWorkspaceRecordings,
} from "./legacyRecordedBackend";
