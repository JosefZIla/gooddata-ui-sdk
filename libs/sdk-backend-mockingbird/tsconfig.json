{
    "compilerOptions": {
        "esModuleInterop": true,
        "target": "es5",
        "module": "commonjs",
        "moduleResolution": "node",
        "lib": ["es6", "dom", "ES2016.Array.Include"],
        "importHelpers": true,
        "declaration": true,
        "declarationMap": true,
        "sourceMap": true,
        "outDir": "./dist",
        "strict": true,
        "noImplicitAny": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "noImplicitReturns": true,
        "baseUrl": "..",
        "paths": {
            "@gooddata/sdk-model": ["sdk-model/src"],
            "@gooddata/sdk-backend-base": ["sdk-backend-base/src"],
            "@gooddata/sdk-backend-spi": ["sdk-backend-spi/src"]
        }
    },
    "exclude": ["node_modules", "dist"]
}
