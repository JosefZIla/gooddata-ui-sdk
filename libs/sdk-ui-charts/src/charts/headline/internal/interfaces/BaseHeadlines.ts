// (C) 2023 GoodData Corporation
import { HeadlineElementType } from "@gooddata/sdk-ui";
import { IHeadlineDataItem } from "./Headlines.js";
import { ComponentType, RefObject } from "react";

export type BaseHeadlineDataItemComponentType = ComponentType<
    IBaseHeadlineDataItemProps & IWithDrillableItemProps & IWithTitleProps
>;

export interface IBaseHeadlineItem {
    data: IHeadlineDataItem;
    baseHeadlineDataItemComponent: BaseHeadlineDataItemComponentType;
    elementType?: HeadlineElementType;
}

export interface IBaseHeadlineData {
    primaryItem: IBaseHeadlineItem;
    secondaryItem?: IBaseHeadlineItem;
    tertiaryItem?: IBaseHeadlineItem;
}

export interface IBaseHeadlineDataItemProps {
    dataItem: IHeadlineDataItem;
}

export interface IWithDrillableItemProps {
    dataItem: IHeadlineDataItem;
    elementType?: HeadlineElementType;
}

export interface IWithTitleProps {
    dataItem: IHeadlineDataItem;
    shouldHideTitle?: boolean;
    titleRef?: RefObject<HTMLDivElement>;
}
