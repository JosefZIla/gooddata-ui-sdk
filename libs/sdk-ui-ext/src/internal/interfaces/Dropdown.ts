// (C) 2019-2021 GoodData Corporation
import { SingleSelectListItemType } from "@gooddata/sdk-ui-kit";

import translations from "../translations/en-US.json";

export interface IDropdownItem {
    title?: string;
    value?: string | number | boolean;
    type?: SingleSelectListItemType;
    icon?: string;
    info?: string;
}

type TranslationKeys = keyof typeof translations;

/**
 * This type narrows down the title to be a key from the translations bundle. Otherwise it provides
 * same properties as the normal dropdown.
 */
export interface ITranslatedDropdownItem extends IDropdownItem {
    title?: TranslationKeys;
}
