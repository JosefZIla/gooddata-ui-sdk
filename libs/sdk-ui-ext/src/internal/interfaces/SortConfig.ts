// (C) 2019-2022 GoodData Corporation

import { Identifier, IAttributeLocatorItem, newMeasureSort } from "@gooddata/sdk-model";
import { MeasureSortSuggestion } from "@gooddata/sdk-ui-kit";

export function newMeasureSortSuggestion(
    identifier: Identifier,
    attributeLocators: IAttributeLocatorItem[] = [],
): MeasureSortSuggestion {
    const {
        measureSortItem: { locators },
    } = newMeasureSort(identifier, "asc", attributeLocators);
    return {
        type: "measureSort",
        locators,
    };
}
