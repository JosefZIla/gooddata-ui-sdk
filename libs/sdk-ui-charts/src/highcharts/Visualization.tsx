// (C) 2007-2018 GoodData Corporation
import { IDataView } from "@gooddata/sdk-backend-spi";
import invariant from "ts-invariant";
import React from "react";
import { IDrillableItem, OnFiredDrillEvent, IHeaderPredicate } from "@gooddata/sdk-ui";
import { ChartTransformation, renderHighCharts } from "./ChartTransformation";
import Highcharts from "./lib";
import { IChartConfig } from "../interfaces";

import { isChartSupported, stringifyChartTypes } from "./chartTypes/_util/common";
import isEqual from "lodash/isEqual";
import isFunction from "lodash/isFunction";
import noop from "lodash/noop";
import omitBy from "lodash/omitBy";

/**
 * @internal
 */
export interface IVisualizationProps {
    height: number;
    width: number;
    config: IChartConfig;
    numericSymbols?: string[];

    dataView: IDataView;
    drillableItems: Array<IDrillableItem | IHeaderPredicate>;
    locale?: string;

    onDrill?: OnFiredDrillEvent;
    afterRender?: () => void;
    onDataTooLarge(): void;
    onNegativeValues(): void;
    onLegendReady(): void;
    pushData?(data: any): void;
}
// TODO: REMOVE VISUALIZATION COMPONENT
/**
 * @internal
 */
export class Visualization extends React.Component<IVisualizationProps> {
    public static defaultProps = {
        locale: "en-US",
        numericSymbols: [] as string[],
        onDrill: (): boolean => true,
        afterRender: noop,
    };

    constructor(props: IVisualizationProps) {
        super(props);

        this.setNumericSymbols(this.props);
    }

    public UNSAFE_componentWillReceiveProps(nextProps: IVisualizationProps): void {
        this.setNumericSymbols(nextProps);
    }

    public shouldComponentUpdate(nextProps: IVisualizationProps): boolean {
        return !isEqual(omitBy(this.props, isFunction), omitBy(nextProps, isFunction));
    }

    public setNumericSymbols(props: IVisualizationProps): void {
        const { numericSymbols } = props;

        if (numericSymbols && numericSymbols.length) {
            Highcharts.setOptions({
                lang: {
                    numericSymbols,
                },
            });
        }
    }

    public render(): React.ReactNode {
        const visType = this.props.config.type;

        if (isChartSupported(visType)) {
            const {
                height,
                width,
                config,
                dataView,
                drillableItems,
                onDrill,
                afterRender,
                onDataTooLarge,
                onNegativeValues,
                onLegendReady,
                locale,
                pushData,
            } = this.props;

            return (
                <ChartTransformation
                    height={height}
                    width={width}
                    config={config}
                    drillableItems={drillableItems}
                    locale={locale}
                    dataView={dataView}
                    afterRender={afterRender}
                    onDrill={onDrill}
                    onDataTooLarge={onDataTooLarge}
                    onNegativeValues={onNegativeValues}
                    onLegendReady={onLegendReady}
                    pushData={pushData}
                    renderer={renderHighCharts}
                />
            );
        }

        invariant(
            false,
            `Unknown visualization type: ${visType}. Supported visualization types: ${stringifyChartTypes()}`,
        );

        return null;
    }
}
