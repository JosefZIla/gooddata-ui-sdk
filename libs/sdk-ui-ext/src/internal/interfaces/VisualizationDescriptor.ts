// (C) 2021 GoodData Corporation
import { ISettings } from "@gooddata/sdk-backend-spi";
import { IInsightDefinition } from "@gooddata/sdk-model";
import { IFluidLayoutDescriptor } from "./LayoutDescriptor";
import { IVisConstruct, IVisualization } from "./Visualization";

/**
 * Factories that create a new instance of pluggable visualization.
 *
 * @alpha
 */
export type PluggableVisualizationFactory = (params: IVisConstruct) => IVisualization;

/**
 * Info about min, max and default in layout units.
 * If no value is provided for min/max/default layout will use its generic defaults for it.
 *
 * @alpha
 */
interface ISizeInfo {
    default?: number;
    min?: number;
    max?: number;
}

/**
 * Info about visualization's min, max and default for width and height.
 *
 * @alpha
 */
export interface IVisualizationSizeInfo {
    width: ISizeInfo;
    height: ISizeInfo;
}

/**
 * @alpha
 * Visualization descriptor
 * Provides access to the visualization's factory and additional details about visualization
 */
export interface IVisualizationDescriptor {
    /**
     * Gets pluggable visualization factory.
     *
     * @alpha
     */
    getFactory(): PluggableVisualizationFactory;

    /**
     * Gets info about size of visualization based on its type, bucket content, layout context and if custom height is enabled.
     * Provided info is used by layout and it defines what will be the default size during render of insight and what will be the limits during resizing of insight.
     *
     * @param insight Current insight definition
     * @param layoutDescriptor layout creating context for visualization
     * @param settings to know if custom heights are allowed
     * @alpha
     */
    getSizeInfo(
        insight: IInsightDefinition,
        layoutDescriptor: IFluidLayoutDescriptor,
        settings: ISettings,
    ): IVisualizationSizeInfo;
}
