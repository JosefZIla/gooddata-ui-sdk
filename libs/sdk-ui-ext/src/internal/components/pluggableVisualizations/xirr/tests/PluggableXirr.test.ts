// (C) 2019-2022 GoodData Corporation
import React from "react";
import ReactDom from "react-dom";
import { dummyBackend } from "@gooddata/sdk-backend-mockingbird";

import { PluggableXirr } from "../PluggableXirr";
import { IVisConstruct, IVisProps } from "../../../../interfaces/Visualization";
import * as referencePointMocks from "../../../../tests/mocks/referencePointMocks";
import * as testMocks from "../../../../tests/mocks/testMocks";
import { IDrillableItem } from "@gooddata/sdk-ui";
import { CoreXirr } from "@gooddata/sdk-ui-charts";

describe("PluggableXirr", () => {
    const defaultProps = {
        projectId: "PROJECTID",
        backend: dummyBackend(),
        element: "body",
        configPanelElement: "invalid",
        visualizationProperties: {},
        callbacks: {
            afterRender: jest.fn(),
            pushData: jest.fn(),
            onLoadingChanged: jest.fn(),
            onError: jest.fn(),
        },
        renderFun: jest.fn(),
    };

    const executionFactory = dummyBackend().workspace("PROJECTID").execution();

    function createComponent(customProps: Partial<IVisConstruct> = {}) {
        return new PluggableXirr({
            ...defaultProps,
            ...customProps,
        });
    }

    describe("init", () => {
        it("should not call pushData during init", () => {
            const pushData = jest.fn();

            createComponent({
                callbacks: {
                    pushData,
                },
            });

            expect(pushData).not.toHaveBeenCalled();
        });
    });

    describe("update", () => {
        const emptyPropertiesMeta = {};

        function getTestOptions(): IVisProps {
            const drillableItems: IDrillableItem[] = [];
            return {
                dimensions: {
                    width: 12,
                    height: 14,
                },
                custom: {
                    drillableItems,
                },
                locale: "en-US",
            };
        }

        it("should not render xirr when dataSource is missing", () => {
            const fakeElement: any = "fake element";
            const reactCreateElementSpy = jest
                .spyOn(React, "createElement")
                .mockImplementation(() => fakeElement);
            const reactRenderSpy = jest.spyOn(ReactDom, "render").mockImplementation(jest.fn());

            const xirr = createComponent();
            const options: IVisProps = getTestOptions();

            xirr.update({ ...options }, testMocks.emptyInsight, emptyPropertiesMeta, executionFactory);

            expect(reactRenderSpy).toHaveBeenCalledTimes(0);

            reactCreateElementSpy.mockReset();
            reactRenderSpy.mockReset();
        });

        it("should render XIRR by react to given element passing down properties", () => {
            const fakeElement: any = "fake element";
            const reactCreateElementSpy = jest
                .spyOn(React, "createElement")
                .mockImplementation(() => fakeElement);
            const mockRenderFun = jest.fn();

            const xirr = createComponent({
                ...defaultProps,
                renderFun: mockRenderFun,
                featureFlags: { enableKDWidgetCustomHeight: false },
            });
            const options: IVisProps = getTestOptions();

            xirr.update(options, testMocks.insightWithSingleMeasure, emptyPropertiesMeta, executionFactory);

            expect(reactCreateElementSpy.mock.calls[0][0]).toBe(CoreXirr);
            expect(mockRenderFun).toHaveBeenCalledWith(
                fakeElement,
                document.querySelector(defaultProps.element),
            );

            reactCreateElementSpy.mockReset();
        });

        it("should render XIRR by react to given element passing down properties when FF enableKDWidgetCustomHeight is set to true", () => {
            const fakeElement: any = "fake element";
            const reactCreateElementSpy = jest
                .spyOn(React, "createElement")
                .mockImplementation(() => fakeElement);
            const mockRenderFun = jest.fn();

            const xirr = createComponent({
                ...defaultProps,
                renderFun: mockRenderFun,
                featureFlags: { enableKDWidgetCustomHeight: true },
            });
            const options: IVisProps = getTestOptions();

            xirr.update(options, testMocks.insightWithSingleMeasure, emptyPropertiesMeta, executionFactory);

            expect(reactCreateElementSpy.mock.calls[0][0]).toBe(CoreXirr);
            expect(mockRenderFun).toHaveBeenCalledWith(
                fakeElement,
                document.querySelector(defaultProps.element),
            );

            reactCreateElementSpy.mockReset();
        });

        it("should correctly set config.disableDrillUnderline from FeatureFlag disableKpiDashboardHeadlineUnderline", () => {
            const fakeElement: any = "fake element";
            const reactCreateElementSpy = jest
                .spyOn(React, "createElement")
                .mockImplementation(() => fakeElement);
            const mockRenderFun = jest.fn();

            const xirr = createComponent({
                featureFlags: {
                    disableKpiDashboardHeadlineUnderline: true,
                },
                renderFun: mockRenderFun,
            });

            const options: IVisProps = getTestOptions();

            xirr.update(options, testMocks.insightWithSingleMeasure, emptyPropertiesMeta, executionFactory);

            expect(reactCreateElementSpy.mock.calls[0][0]).toBe(CoreXirr);

            reactCreateElementSpy.mockReset();
        });
    });

    describe("getExtendedReferencePoint", () => {
        it("should return proper extended reference point", async () => {
            const extendedReferencePoint = await createComponent().getExtendedReferencePoint(
                referencePointMocks.measuresAndDateReferencePoint,
            );

            expect(extendedReferencePoint).toMatchSnapshot();
        });

        it("should correctly process empty reference point", async () => {
            const headline = createComponent();

            const extendedReferencePoint = await headline.getExtendedReferencePoint(
                referencePointMocks.emptyReferencePoint,
            );

            expect(extendedReferencePoint).toMatchSnapshot();
        });
    });
});
