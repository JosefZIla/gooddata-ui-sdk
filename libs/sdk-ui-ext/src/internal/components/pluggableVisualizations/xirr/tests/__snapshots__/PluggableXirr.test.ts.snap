// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluggableXirr getExtendedReferencePoint should correctly process empty reference point 1`] = `
Object {
  "buckets": Array [
    Object {
      "items": Array [],
      "localIdentifier": "measures",
    },
    Object {
      "items": Array [],
      "localIdentifier": "attribute",
    },
  ],
  "filters": Object {
    "items": Array [],
    "localIdentifier": "filters",
  },
  "properties": Object {},
  "uiConfig": Object {
    "buckets": Object {
      "attribute": Object {
        "accepts": Array [
          "date",
        ],
        "allowsReordering": false,
        "allowsSwapping": true,
        "canAddItems": true,
        "enabled": true,
        "isShowInPercentEnabled": false,
        "itemsLimit": 1,
        "itemsLimitByType": Object {
          "date": 1,
        },
        "title": "Date attribute",
      },
      "filters": Object {
        "accepts": Array [
          "attribute",
          "date",
        ],
        "allowsReordering": false,
        "enabled": true,
        "isShowInPercentEnabled": false,
        "itemsLimit": 20,
        "itemsLimitByType": Object {
          "date": 1,
        },
      },
      "measures": Object {
        "accepts": Array [
          "metric",
          "fact",
        ],
        "allowsDuplicateItems": true,
        "allowsReordering": true,
        "allowsSwapping": true,
        "canAddItems": true,
        "enabled": true,
        "isShowInPercentEnabled": false,
        "isShowInPercentVisible": false,
        "itemsLimit": 1,
        "title": "Measure",
      },
    },
    "customError": Object {
      "heading": "Incomplete configuration",
      "text": "Make sure you have selected one measure and one date attribute",
    },
    "exportConfig": Object {
      "supported": false,
    },
    "openAsReport": Object {
      "supported": false,
    },
    "recommendations": Object {},
    "supportedOverTimeComparisonTypes": Array [],
  },
}
`;

exports[`PluggableXirr getExtendedReferencePoint should return proper extended reference point 1`] = `
Object {
  "buckets": Array [
    Object {
      "items": Array [
        Object {
          "aggregation": null,
          "attribute": "aazb6kroa3iC",
          "localIdentifier": "m1",
          "showInPercent": null,
          "showOnSecondaryAxis": null,
          "type": "metric",
        },
      ],
      "localIdentifier": "measures",
    },
    Object {
      "items": Array [
        Object {
          "aggregation": null,
          "attribute": "attr.datedataset",
          "dateDatasetRef": Object {
            "identifier": "ds1",
          },
          "localIdentifier": "a1",
          "showInPercent": null,
          "type": "date",
        },
      ],
      "localIdentifier": "attribute",
    },
  ],
  "filters": Object {
    "items": Array [],
    "localIdentifier": "filters",
  },
  "properties": Object {},
  "uiConfig": Object {
    "buckets": Object {
      "attribute": Object {
        "accepts": Array [
          "date",
        ],
        "allowsReordering": false,
        "allowsSwapping": true,
        "canAddItems": false,
        "enabled": true,
        "isShowInPercentEnabled": false,
        "itemsLimit": 1,
        "itemsLimitByType": Object {
          "date": 1,
        },
        "title": "Date attribute",
      },
      "filters": Object {
        "accepts": Array [
          "attribute",
          "date",
        ],
        "allowsReordering": false,
        "enabled": true,
        "isShowInPercentEnabled": false,
        "itemsLimit": 20,
        "itemsLimitByType": Object {
          "date": 1,
        },
      },
      "measures": Object {
        "accepts": Array [
          "metric",
          "fact",
        ],
        "allowsDuplicateItems": true,
        "allowsReordering": true,
        "allowsSwapping": true,
        "canAddItems": false,
        "enabled": true,
        "isShowInPercentEnabled": false,
        "isShowInPercentVisible": false,
        "itemsLimit": 1,
        "title": "Measure",
      },
    },
    "customError": undefined,
    "exportConfig": Object {
      "supported": false,
    },
    "openAsReport": Object {
      "supported": false,
    },
    "recommendations": Object {},
    "supportedOverTimeComparisonTypes": Array [],
  },
}
`;
